### PR TITLE
Instance weights

### PR DESCRIPTION
**Pull request by tdhd**
A follow up on #188.

I renamed some variables and added a standardization for the newly added `instance_weight` parameter. Will add a unifying test case for class- and instance weights later.

### CODE REVIEW

File: keras/models.py [hunk 1]
```diff
@@ -41,8 +41,19 @@ def slice_X(X, start=None, stop=None):
...
 
-def calculate_class_weights(Y, class_weight):
-    if isinstance(class_weight, dict):
+def standardize_instance_weights(instance_weight):
+    if isinstance(instance_weight, list):
+        w = np.array(instance_weight)
+    elif isinstance(instance_weight, np.ndarray):
+        w = instance_weight
+    else:
+        w = None
```
Comment (fchollet) [inline keras/models.py]: Is `None` viable? If not, shouldn't we throw an exception instead of returning `None`?

**Comment by fchollet** [comment]
Looking good. I'd personally favor `sample_weight` instead of `instance_weight`, to match the [sklearn conventions](http://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC.fit) (same as class_weight).

**Comment by tdhd** [comment]
Ok, I renamed the parameter.

To return `None` in the standardization of the sample weights is required by `calculate_loss_weights`. Also for the inner batch loop in `fit` I needed to add a check for `None` because of the sample weight slicing.

In the test case I am using the sample weights to simulate class weights. Although I do that, the predicted results on the test set are slightly different. I suspect it is a bug in the test file which is likely caused by the messy code. :)

I will clean it up when I have time again.

**Comment by fchollet** [comment]
> I will look into it when I have a little more time on my hands.

Ok, I'll wait for you to clear this up : )

About the implementation: why not standardize the approach and only use `sample_weight` to scale the loss? If provided, `class_weight` would be used to generate a `sample_weight` at the beginning of `fit`. 

Something like:
```python
# at start of fit 
if sample_weight:
   sample_weight = standardize_weights(sample_weight)
elif class_weight:
   sample_weight = calculate_loss_weights(Y, class_weight)
else:
   sample_weight = np.ones((Y.shape[0]))
```

Which can be abstracted in:
```python
sample_weight = standardize_weights(Y, sample_weight, class_weight)
```

Then in the batch loop simply:
```python
w = sample_weight[batch_ids]
```
I think it could make the process simpler / clearer. Unless there is a reason not to do it?

**Comment by tdhd** [comment]
> I think it could make the process simpler / clearer. Unless there is a reason not to do it?

Completely agreed.

**Comment by fchollet** [comment]
I rewrote the tests. Loss weighting does not work well with hinge and squared hinge, but that does not appear to be an implementation issue. I excluded these two losses from the tests.

**Comment by fchollet** [comment]
FYI I reverted the entire loss weighting feature because it caused a bug in sequence to sequence training. 

Specifically, any network of this type would crash at training time:
```python
model = Sequential()
model.add(SimpleRNN(n_in, n_hidden, return_sequences=True))
model.compile(loss='mse', optimizer='rmsprop')
```

The cause of the issue is not clear. I believe it has to do with the modifications made to the objective functions. 

We'll reintroduce loss weighting when that is cleared up.

**Comment by tdhd** [comment]
Oh damn it, sorry I didn't test RNNs with the modifications. I hope we see class- and instance weights soon to be back though.

**Comment by fchollet** [comment]
It's back now, though the implementation is somewhat different.

**Comment by tdhd** [comment]
Awesome.
